File: Java/LinkedList.java
import java.util.Scanner;

class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    private Node head;

    void addElementToEnd(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    void addElementToBeginning(int data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }

    void addElementAfterValue(int value, int newValue) {
        Node current = head;
        while (current != null) {
            if (current.data == value) {
                Node newNode = new Node(newValue);
                newNode.next = current.next;
                current.next = newNode;
                break;
            }
            current = current.next;
        }
    }

    void addElementBeforeValue(int value, int newValue) {
        if (head != null && head.data == value) {
            addElementToBeginning(newValue);
            return;
        }

        Node prev = null;
        Node current = head;
        while (current != null) {
            if (current.data == value) {
                Node newNode = new Node(newValue);
                prev.next = newNode;
                newNode.next = current;
                break;
            }
            prev = current;
            current = current.next;
        }
    }

    void removeElementFromBeginning() {
        if (head != null) {
            head = head.next;
        }
    }

    void removeElementFromEnd() {
        if (head != null) {
            if (head.next == null) {
                head = null;
                return;
            }
            Node prev = null;
            Node current = head;
            while (current.next != null) {
                prev = current;
                current = current.next;
            }
            if (prev != null) {
                prev.next = null;
            }
        }
    }

    void removeElement(int value) {
        if (head != null && head.data == value) {
            removeElementFromBeginning();
            return;
        }

        Node prev = null;
        Node current = head;
        while (current != null) {
            if (current.data == value) {
                prev.next = current.next;
                break;
            }
            prev = current;
            current = current.next;
        }
    }

    boolean findElement(int value) {
        Node current = head;
        while (current != null) {
            if (current.data == value) {
                return true;
            }
            current = current.next;
        }
        return false;
    }

    void reverseList() {
        Node prev = null;
        Node current = head;
        Node next;
        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    void clearList() {
        head = null;
    }

    void mergeLists(LinkedList list2) {
        Node current = head;
        if (current != null) {
            while (current.next != null) {
                current = current.next;
            }
            current.next = list2.head;
        } else {
            head = list2.head;
        }
    }

    LinkedList copyList() {
        LinkedList newList = new LinkedList();
        Node current = head;
        while (current != null) {
            newList.addElementToEnd(current.data);
            current = current.next;
        }
        return newList;
    }

    void printList() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }
}

public class Main {
    public static void main(String[] args) {
        LinkedList linkedList = new LinkedList();
        Scanner scanner = new Scanner(System.in);
        int option;

        do {
            System.out.println("Options");
            System.out.println("1. Add element to the end of the list");
            System.out.println("2. Add element to the beginning of the list");
            System.out.println("3. Add element after the specified value");
            System.out.println("4. Add element before the specified value");
            System.out.println("5. Remove element from the beginning of the list");
            System.out.println("6. Remove element from the end of the list");
            System.out.println("7. Remove element from the list");
            System.out.println("8. Find element in the list");
            System.out.println("9. Print the list");
            System.out.println("10. Reverse the list");
            System.out.println("11. Clear the list");
            System.out.println("12. Merge two lists");
            System.out.println("13. Copy the list");
            System.out.println("14. Exit");
            System.out.print("Enter your choice: ");
            option = scanner.nextInt();

            switch (option) {
                case 1:
                    System.out.print("Enter the element to add at the end: ");
                    int data = scanner.nextInt();
                    linkedList.addElementToEnd(data);
                    break;
                case 2:
                    System.out.print("Enter the element to add at the beginning: ");
                    int beginningData = scanner.nextInt();
                    linkedList.addElementToBeginning(beginningData);
                    break;
                case 3:
                    System.out.print("Enter the value after which to add: ");
                    int value = scanner.nextInt();
                    System.out.print("Enter the new element: ");
                    int newData = scanner.nextInt();
                    linkedList.addElementAfterValue(value, newData);
                    break;
                case 4:
                    System.out.print("Enter the value before which to add: ");
                    int beforeValue = scanner.nextInt();
                    System.out.print("Enter the new element: ");
                    int newBeforeData = scanner.nextInt();
                    linkedList.addElementBeforeValue(beforeValue, newBeforeData);
                    break;
                case 5:
                    linkedList.removeElementFromBeginning();
                    break;
                case 6:
                    linkedList.removeElementFromEnd();
                    break;
                case 7:
                    System.out.print("Enter the element to remove: ");
                    int removeData = scanner.nextInt();
                    linkedList.removeElement(removeData);
                    break;
                case 8:
                    System.out.print("Enter the element to find: ");
                    int findData = scanner.nextInt();
                    if (linkedList.findElement(findData)) {
                        System.out.println(findData + " found in the list");
                    } else {
                        System.out.println(findData + " not found in the list");
                    }
                    break;
                case 9:
                    linkedList.printList();
                    break;
                case 10:
                    linkedList.reverseList();
                    break;
                case 11:
                    linkedList.clearList();
                    break;
                case 12:
                    LinkedList list2 = new LinkedList();
                    System.out.print("Enter elements of second list separated by spaces: ");
                    scanner.nextLine(); // Consumir el carácter de nueva línea restante
                    String input = scanner.nextLine();
                    String[] elements = input.split(" ");
                    for (String element : elements) {
                        list2.addElementToEnd(Integer.parseInt(element));
                        linkedList.mergeLists(list2);
                        break;
                    }
                case 13:
                    LinkedList copyList = linkedList.copyList();
                    System.out.println("Copied list:");
                    copyList.printList();
                    break;
                case 14:
                    System.out.println("Exiting...");
                    break;
                default:
                    System.out.println("Invalid option");
                }
                // Consumir el carácter de nueva línea restante
                scanner.nextLine();
            } while (option != 14);
        }
    }
    

File: Python/LinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def add_element_to_end(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node

    def add_element_to_beginning(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def add_element_after_value(self, value, new_value):
        current = self.head
        while current:
            if current.data == value:
                new_node = Node(new_value)
                new_node.next = current.next
                current.next = new_node
                break
            current = current.next

    def add_element_before_value(self, value, new_value):
        if self.head.data == value:
            self.add_element_to_beginning(new_value)
            return
        prev = None
        current = self.head
        while current:
            if current.data == value:
                new_node = Node(new_value)
                prev.next = new_node
                new_node.next = current
                break
            prev = current
            current = current.next

    def remove_element_from_beginning(self):
        if self.head:
            self.head = self.head.next

    def remove_element_from_end(self):
        if not self.head:
            return
        if not self.head.next:
            self.head = None
            return
        prev = None
        current = self.head
        while current.next:
            prev = current
            current = current.next
        prev.next = None

    def remove_element(self, value):
        if self.head.data == value:
            self.remove_element_from_beginning()
            return
        prev = None
        current = self.head
        while current:
            if current.data == value:
                prev.next = current.next
                break
            prev = current
            current = current.next

    def find_element(self, value):
        current = self.head
        while current:
            if current.data == value:
                return True
            current = current.next
        return False

    def reverse_list(self):
        prev = None
        current = self.head
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        self.head = prev

    def clear_list(self):
        self.head = None

    def merge_lists(self, list2):
        current = self.head
        while current.next:
            current = current.next
        current.next = list2.head

    def copy_list(self):
        new_list = LinkedList()
        current = self.head
        while current:
            new_list.add_element_to_end(current.data)
            current = current.next
        return new_list

    def print_list(self):
        current = self.head
        while current:
            print(f"{current.data} -> ", end="")
            current = current.next
        print("null")

if __name__ == "__main__":
    linked_list = LinkedList()
    option = 0

    while option != 14:
        print("Options")
        print("1. Add element to the end of the list")
        print("2. Add element to the beginning of the list")
        print("3. Add element after the specified value")
        print("4. Add element before the specified value")
        print("5. Remove element from the beginning of the list")
        print("6. Remove element from the end of the list")
        print("7. Remove element from the list")
        print("8. Find element in the list")
        print("9. Print the list")
        print("10. Reverse the list")
        print("11. Clear the list")
        print("12. Merge two lists")
        print("13. Copy the list")
        print("14. Exit")
        option = int(input("Enter your choice: "))

        if option == 1:
            data = int(input("Enter the element to add at the end: "))
            linked_list.add_element_to_end(data)
        elif option == 2:
            data = int(input("Enter the element to add at the beginning: "))
            linked_list.add_element_to_beginning(data)
        elif option == 3:
            value = int(input("Enter the value after which to add: "))
            new_data = int(input("Enter the new element: "))
            linked_list.add_element_after_value(value, new_data)
        elif option == 4:
            value = int(input("Enter the value before which to add: "))
            new_data = int(input("Enter the new element: "))
            linked_list.add_element_before_value(value, new_data)
        elif option == 5:
            linked_list.remove_element_from_beginning()
        elif option == 6:
            linked_list.remove_element_from_end()
        elif option == 7:
            data = int(input("Enter the element to remove: "))
            linked_list.remove_element(data)
        elif option == 8:
            data = int(input("Enter the element to find: "))
            if linked_list.find_element(data):
                print(f"{data} found in the list")
            else:
                print(f"{data} not found in the list")
        elif option == 9:
            linked_list.print_list()
        elif option == 10:
            linked_list.reverse_list()
        elif option == 11:
            linked_list.clear_list()
        elif option == 12:
            list2 = LinkedList()
            elements = input("Enter elements of second list separated by spaces: ").split()
            for element in elements:
                list2.add_element_to_end(int(element))
            linked_list.merge_lists(list2)
        elif option == 13:
            copy_list = linked_list.copy_list()
            print("Copied list:")
            copy_list.print_list()
        elif option == 14:
            print("Exiting...")
        else:
            print("Invalid option")
